### PR TITLE
feat: add Pi Coding Agent observability support

### PR DESCRIPTION
**Pull request by Goclis**
## Summary

- add first-class support for Pi Coding Agent using the canonical `pi-coding-agent` ID
- inject a repository-maintained Pi Extension that captures LLM request/response and tool lifecycle events
- collect and normalize the generated JSONL events through a dedicated `pi-coding-agent-log` input
- integrate deployment, watchdog repair, uninstall cleanup, content-capture policy, documentation, and agent metadata across Linux/macOS and Windows

## Implementation details

- deploys the Extension through Pi's `~/.pi/agent/settings.json` `extensions` configuration
- emits `gen_ai` attributes including request/response model, response ID, token usage, finish reasons, tool definitions, and tool results
- respects `captureMessageContent=false`, including server-side removal of message content, system instructions, and tool definitions
- uses restrictive permissions for newly created settings, log directories, and log files
- consistently reports `gen_ai.agent.type=pi-coding-agent` and `gen_ai.system=pi`

## Validation

- `npm run typecheck`
- `npm run build`
- `npm test -- --run` — 153 test files passed; 1685 tests passed, 2 skipped
- packaged and locally upgraded Pilot to `v1.1.18`; collector, updater, Extension deployment, and `pi-coding-agent-log` registration verified
- real Pi Coding Agent session verified through raw Extension JSONL and normalized Pilot output for `llm.request` and `llm.response`

### CODE REVIEW

File: assets/plugins/pi-coding-agent/index.mjs [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+function canonicalMessage(message) {
+  if (!message || typeof message !== 'object') return null;
+
+  if (message.role === 'toolResult') {
+    return {
+      role: 'tool',
+      parts: [{
+        type: 'tool_call_response',
+        id: message.toolCallId || null,
+        response: contentParts(message.content),
```
Comment (linrunqi08) [inline assets/plugins/pi-coding-agent/index.mjs]: **[Medium]** `tool_call_response.response` 类型不一致：`toolResult` 分支输出**数组**（`contentParts(...)`），而同文件 `bashExecution` 分支(L237)与 OpenCode 插件(`plugin.mjs:315`)均输出**字符串**。

**影响:** 同一 canonical 字段 `response` 在不同 agent、甚至本文件内类型不一致；下游 schema 校验/UI 渲染若按字符串处理，将错渲或丢弃 Pi 工具结果。
**建议:** 输出字符串，与 bashExecution/OpenCode 约定对齐：`response: typeof message.content === "string" ? truncate(message.content) : truncate(safeStringify(message.content))`。

**Review by ralf0131 — COMMENTED** [review]
﻿### Summary
Comprehensive, well-structured implementation of Pi Coding Agent observability — adds a new agent definition, a dependency-free fail-open extension plugin, a log input collector, and extends `PluginInjectStrategy` with `configKey`/`createIfMissing` support. The code follows existing patterns, respects content-capture privacy settings, and has thorough test coverage across all new and modified components.

⚠️ **CLA not signed** — this PR cannot be approved/merged until the contributor signs the CLA. Please check the CLA reminder above.

### Findings
- **[Info]** `src/deployment/plugin-inject-strategy.ts:210` — When `createIfMissing` is true, only `configPaths[0]` is used for file creation. If the first path is in a read-only location, the strategy will fail without trying alternative paths. Consider iterating through `configPaths` for the create step as well.
- **[Info]** `assets/plugins/pi-coding-agent/index.mjs:105` — `resolveDataDir()` falls back to `installedDataDir` (two levels up from the plugin file). This is correct for standard deployment, but a brief inline comment would help clarify the expected directory layout.

### Positive Highlights
- **Fail-open design** — `safeHandler` wraps all event handlers so telemetry failures never interrupt the agent turn or tool execution
- **Privacy-aware** — `captureMessageContent` is respected throughout, including server-side removal of message content, system instructions, and tool definitions
- **Backward-compatible** — `configKey` and `createIfMissing` are optional, defaulting to existing behavior
- **Race-safe file creation** — `createIfMissing` uses `flag: 'wx'` (exclusive) with proper `EEXIST` handling
- **Comprehensive tests** — extension behavior, input registration, plugin-inject strategy, uninstall cleanup (sh + ps1), config-loader, orchestrator watchdog, content policy, and agent-system-map tests
- **Cross-platform installer cleanup** — both `.sh` and `.ps1` handle Pi extension removal on uninstall

---
*Automated review by github-manager-bot*

File: src/deployment/plugin-inject-strategy.ts [hunk 2]
```diff
@@ -197,15 +198,35 @@ export class PluginInjectStrategy implements DeployStrategy {
...
+  private async findConfigFile(
+    config: PluginInjectConfig,
+    createIfMissing: boolean,
+  ): Promise<string | null> {
     for (const p of config.configPaths) {
       const resolved = resolveHome(p);
       if (await fileExists(resolved)) return resolved;
     }
+
+    if (createIfMissing && config.configPaths.length > 0) {
```
Comment (ralf0131) [inline src/deployment/plugin-inject-strategy.ts:210]: **[Info]** When `createIfMissing` is true, only `configPaths[0]` is used for file creation. If the first path is in a read-only location, the strategy will fail without trying alternative paths. Consider iterating through `configPaths` for the create step as well, or documenting that the first path must be writable.

File: assets/plugins/pi-coding-agent/index.mjs [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+function ensureLogDir() {
+  fs.mkdirSync(resolveLogDir(), { recursive: true, mode: 0o700 });
+}
+
+function writeError(source, error) {
+  try {
+    ensureLogDir();
+    const file = path.join(resolveLogDir(), `${AGENT_TYPE}-error-${todayStamp()}.log`);
+    fs.appendFileSync(
+      file,
```
Comment (ralf0131) [inline assets/plugins/pi-coding-agent/index.mjs]: **[Info]** `resolveDataDir()` falls back to `installedDataDir` (two levels up from the plugin file). This is correct for standard deployment, but worth a brief inline comment to clarify the expected directory layout for anyone reading the extension standalone.

File: src/core/orchestrator.ts [hunk 4]
```diff
@@ -1034,6 +1037,27 @@ export class Orchestrator extends EventEmitter {
       }),
     );
 
+    // --- Pi Coding Agent Log (Pi extension JSONL) ---
+    const piCodingAgentLogDir = path.join(this.dataDir, 'logs', 'pi-coding-agent');
+    await ensureDir(piCodingAgentLogDir);
```
Comment (linrunqi08) [inline src/core/orchestrator.ts]: **[Medium]** 日志目录 `0o700` 意图被静默破坏。此处 `ensureDir(piCodingAgentLogDir)` → `fs-utils.ts:135` 的 `mkdir(recursive:true)` **不带 mode**，目录实际为 `0o755`；extension 的 `mkdirSync(..., {mode:0o700})`(index.mjs:97) 对**已存在目录是 no-op**，守护进程先行创建后 `0o700` 失效。

**影响:** 存放捕获消息/system prompt/工具参数的目录对本机其他用户可枚举（文件 `0o600` 内容不可读，但目录列举与会话日期泄露，defense-in-depth 失效）。
**建议:** 给 `ensureDir` 增加可选 `mode` 并对该目录传 `0o700`，或创建后显式 `chmod 0o700`。

File: assets/plugins/pi-coding-agent/index.mjs [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+  } catch {
+    // Telemetry errors are intentionally ignored.
+  }
+}
+
+function writeRecord(record) {
+  try {
+    ensureLogDir();
+    const file = path.join(resolveLogDir(), `${AGENT_TYPE}-${todayStamp()}.jsonl`);
+    fs.appendFileSync(file, `${safeStringify(record)}\n`, { encoding: 'utf8', mode: 0o600 });
```
Comment (linrunqi08) [inline assets/plugins/pi-coding-agent/index.mjs]: **[Medium]** `writeRecord`/`writeError` 在**每个事件**上先 `ensureLogDir()`（`mkdirSync` syscall，即使目录已存在）再 `appendFileSync`，全部同步、运行在 Pi 进程主线程。

**影响:** 每条 `llm.request/response`、`tool.call/result` 多一次冗余 `mkdir` + 一次同步阻塞写；慢速/NFS 数据目录下给 agent turn 注入同步 IO 延迟，与文件头部“telemetry 绝不打断 agent turn”目标相悖。
**建议:** 首次成功创建后缓存 `dirReady` 标志，仅在 append 报 `ENOENT` 时重建；如可行改异步/队列写。

**Comment by linrunqi08** [comment]
## 🔍 Code Review Summary

| Severity | Count |
|----------|-------|
| Critical | 0 |
| High     | 0 |
| Medium   | 3 |
| Low      | 6 |

### Lifecycle Verdict

| Check | Result |
|-------|--------|
| 资源释放 | PASS |
| 死锁/卡死风险 | PASS (慢盘同步 IO 延迟隐患，非 hang) |
| 状态恢复正确性 | PASS |

### 总体结论

**未发现阻断合入的问题（无 Critical / High）。** 变更与既有 plugin-inject（OpenCode）范式高度一致，部署/采集/自愈/卸载链路自洽，`captureMessageContent` 有服务端二次防御，token 口径与 `claude-code-hook-processor` 一致。

3 个 Medium 建议合入前处理：
1. **[Medium] `tool_call_response.response` 类型不一致** (`assets/plugins/pi-coding-agent/index.mjs:216`)：`toolResult` 分支输出数组，而同文件 `bashExecution`(L237) 与 OpenCode 插件均输出字符串 → 下游 schema/渲染可能错渲或丢弃 Pi 工具结果。
2. **[Medium] 日志目录 `0o700` 被静默破坏** (`src/core/orchestrator.ts:1042`)：`ensureDir` 不带 mode 先建目录为 `0o755`，extension 的 `mkdirSync(mode:0o700)` 对已存在目录 no-op → 目录可被本机其他用户枚举，defense-in-depth 失效。
3. **[Medium] 每事件同步 `mkdirSync`+`appendFileSync`** (`assets/plugins/pi-coding-agent/index.mjs:118`)：Pi 主线程热路径注入同步 IO，慢盘/NFS 下会拖慢 agent turn。建议缓存 `dirReady` 并评估异步写。

Low（详见报告）：`error.message` 未受 `captureMessageContent` 门控（隐私）；`turn_start` 兜底 turnId 未回写致 step↔turn 关联偏差；日志无大小滚动；多进程共享文件写入撕裂；卸载脚本缺 JSONC 回退 + 非原子写；`toSerializable` 的 `__proto__` own key 丢弃（不可利用）。

### 主要测试缺口

- error 记录路径（`stopReason==='error'` / `tool isError`）未测 —— 核心可观测路径。
- `canonicalMessage` 角色变体（toolResult / bashExecution / custom / image）、`context` 去重门控、无匹配 start 的 duration 分支未测。
- 卸载 `node -e` 清理仅字符串匹配脚本文本，未对样例 `settings.json` 实际执行。

### Highlights（正向实践）

- extension dependency-free + fail-open，telemetry 失败不打断 agent。
- 正确复用 `BaseHookInput`/`transformHookRecord`，且正确**不**启用 `coldStartKeepLastTurnOnly`（独立进程写入语义）。
- 隐私双重防御（extension 端不采集 + 服务端 content-policy 剥离 `system_instructions`/`tool.definitions`）。
- `toSerializable` 全套上限 + 循环引用防御；文档（含采集矩阵中英文）更新完整。

---
*评审报告详见: `code-review/pr-154/final-report.md`*
*Generated by LoongSuite-Pilot Code Review Agent*

**Review by ralf0131 — APPROVED** [review]
### Summary
Re-reviewing for approval: CLA is now signed (was pending at previous review). The code review from the previous COMMENT remains positive — comprehensive, well-structured implementation of Pi Coding Agent observability with fail-open design, privacy-aware content capture, backward-compatible optional fields, race-safe file creation, and thorough cross-platform test coverage.

The two [Info] suggestions from the previous review (iterate `configPaths` for create step; inline comment for `resolveDataDir` layout) are non-blocking improvements — safe to approve.

Approving now that CLA is signed. CI workflow run approved for first-time contributor.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
Re-reviewing commit `8055f05` — the contributor addressed both [Info] suggestions from the previous review and added several robustness improvements. The implementation remains comprehensive, well-structured, and well-tested.

### Changes in the new commit
- **[Addressed]** `index.mjs:25` — Added inline comment clarifying the installed layout (`$PILOT_DATA/plugins/pi-coding-agent/index.mjs`), directly resolving the previous [Info] suggestion about `resolveDataDir()` documentation.
- **[Improved]** `index.mjs` — Introduced `appendLogFile()` with ENOENT retry logic and a `logDirReady` cache flag, making log writing resilient to runtime directory removal without redundant `mkdirSync` calls.
- **[Improved]** `index.mjs` — Added `fs.chmodSync(dir, 0o700)` on non-Windows platforms to tighten permissions on pre-existing log directories.
- **[Improved]** `index.mjs` — Added `toolResultResponse()` to serialize tool results as plain text, preventing image/binary content from being embedded in JSONL logs (privacy enhancement).
- **[Improved]** `index.mjs` — Made error message inclusion conditional on `captureContent`, preventing potential prompt leakage in error logs when content capture is disabled.
- **[Improved]** `orchestrator.ts` — Uses `ensurePiCodingAgentLogDir()` which creates the directory with `0o700` and tightens existing permissions.
- **[Improved]** `pi-coding-agent-log-input.ts` — Added `onStart()` to ensure the log directory exists before reading, with proper permission handling.
- **[Tests]** Added 143 lines of new tests covering retry logic, platform-specific permission handling, tool result serialization, and turn/step ID correlation.

### Verdict
LGTM — the new commit addresses the previous review suggestions and adds meaningful robustness improvements. All changes are well-tested and follow existing patterns. Approving on the new commit.

---
*Automated review by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
Re-reviewing commit `8f05411` — a test-only commit that adds edge-case coverage for the Pi Coding Agent extension. The previous two reviews (on commits `f4935e4` and `8055f05`) were APPROVED.

The new tests cover:
- **Message type conversion**: bash execution and custom messages correctly mapped to canonical `gen_ai.input.messages` (tool role with `is_error` flag, user role with text content)
- **LLM request deduplication**: multiple context emits within a single step produce only one LLM request; step counter resets correctly across turns
- **Tool failure recording**: tool execution failures capture status, error type, result, and duration (computed from `tool_execution_start`/`tool_execution_end` timestamps)
- **Defensive handling**: tool duration omitted when no matching `tool_execution_start` was observed

Tests are well-structured with deterministic timing via `vi.setSystemTime`, clear assertions, and good edge-case coverage. CLA is signed. LGTM.

---
*Automated review by github-manager-bot*

File: assets/plugins/pi-coding-agent/index.mjs (hunk 1 above)
Comment (Goclis) [reply assets/plugins/pi-coding-agent/index.mjs]: Fixed in commit `8055f05`. Pi tool results now go through `toolResultResponse()` and always emit a string response. Plain text remains plain text; mixed content is serialized from sanitized canonical parts so image payload data is not embedded. Covered by the tool-result serialization test in `pi-coding-agent-extension.test.mjs`.

File: src/deployment/plugin-inject-strategy.ts (hunk 2 above)
Comment (Goclis) [reply src/deployment/plugin-inject-strategy.ts:210]: Thanks for the suggestion. We decided not to change this behavior in this PR. `createIfMissing` is explicitly documented as creating the first configured path, and Pi Coding Agent declares exactly one config path (`~/.pi/agent/settings.json`). Trying alternative paths would change the generic `PluginInjectStrategy` contract, so that should be considered separately if a future agent needs writable fallback paths.

File: assets/plugins/pi-coding-agent/index.mjs (hunk 3 above)
Comment (Goclis) [reply assets/plugins/pi-coding-agent/index.mjs]: Fixed in commit `8055f05`. Added an inline comment documenting the expected installed layout: `$PILOT_DATA/plugins/pi-coding-agent/index.mjs`, which explains why resolving two parent directories yields the Pilot data directory.

File: src/core/orchestrator.ts (hunk 4 above)
Comment (Goclis) [reply src/core/orchestrator.ts]: Fixed in commit `8055f05`. The orchestrator and Pi input now use `ensurePiCodingAgentLogDir()`, which creates the directory with mode `0700` and explicitly chmods an existing directory on non-Windows platforms. The Extension also tightens an existing directory before writing. Regression tests cover both the Extension and Input startup paths.

File: assets/plugins/pi-coding-agent/index.mjs (hunk 5 above)
Comment (Goclis) [reply assets/plugins/pi-coding-agent/index.mjs]: Addressed in commit `8055f05`. A process-level `logDirReady` cache removes the per-event `mkdirSync`, and `appendLogFile()` resets the cache and recreates the directory on `ENOENT`. Regression tests cover both one-time directory creation and runtime directory removal. We intentionally retained ordered synchronous append for this PR; moving writes to an asynchronous queue requires shutdown draining and ordering guarantees and is better handled as a broader plugin-framework performance change.

**Comment by ralf0131** [comment]
👋 Pinging maintainers for merge follow-up on this **first-time contributor** PR.

Status: **approved** · **CI passing** (build-and-test ×3, license/cla ✓ signed) · **mergeable** (no conflicts). It has been awaiting merge for ~28h since the latest approval (commit `8f05411`, approved 2026-07-16 13:37 UTC).

Timely merges for first-time contributor PRs help retain new contributors — cc @Monsooooon, would appreciate a merge when convenient. 🙏

---
*Automated follow-up by github-manager-bot*

**Review by ralf0131 — APPROVED** [review]
### Summary
Adds first-class Pi Coding Agent observability support — agent definition, repository-maintained Pi Extension (index.mjs), dedicated `pi-coding-agent-log` input, deployment strategy enhancements, content policy additions, and system map mapping. Overall verdict: **LGTM, looks good to approve.**

### Findings
No critical or blocking issues found. The implementation is clean, well-structured, and follows existing agent integration patterns (cursor, claude-code, codex, etc.).

### Highlights
- **Security**: Restrictive file permissions (0o700 dirs, 0o600 files) for newly created config and log directories; `flag: 'wx'` with EEXIST catch for race-safe config creation.
- **Privacy**: `gen_ai.system_instructions` and `gen_ai.tool.definitions` correctly added to content policy filter set; `captureMessageContent=false` properly suppresses message content, tool definitions, and system instructions.
- **Fail-open design**: Extension errors are caught and logged without interrupting agent turns or tool execution.
- **Serialization safety**: `toSerializable` with depth limit (8), array cap (100), key cap (100), string truncation (64KB), and circular reference detection via WeakSet.
- **Backward compatibility**: `resolvePluginSpecPath` adds absolute path support alongside existing `file://` handling; `resolvePluginKey` falls back to auto-detection when `configKey` is unset.

### Suggestions
- Consider documenting the expected Pi extension lifecycle event names (`session_start`, `before_agent_start`, `turn_start`, `context`, `message_end`, `tool_execution_start`/`end`) in the agent-onboarding guide for future contributors.
- The `requestEmitted` flag in `context` handler prevents duplicate LLM request events — worth a brief inline comment explaining the deduplication rationale for maintainability.

---
*Automated review by github-manager-bot*

**Comment by fangxiu-wf** [comment]
Thanks for this — it's a well-structured contribution, and the Pi extension in particular is nicely done: dependency-free and fail-open (`safeHandler` around every handler), bounded serialization (depth/array/object/string caps + circular guard), restrictive file perms (0700 dir / 0600 files), and a clean double-layer `captureMessageContent` gate (client-side in the extension + server-side via `agent-content-policy`). Deployment changes (`createIfMissing` with `wx` exclusive-create, `configKey` support) are backward-compatible and race-safe. Test and docs coverage is thorough.

**E2E result (dual-checked against a ~12-min live Pi run):** the pipeline is fully functional — 3 traces / 85 spans, ENTRY→AGENT→STEP→LLM/TOOL tree complete and correct, all cross-field semantic checks pass (session/user/agent consistency, token summation, tool↔LLM matching), install-from-tarball + collector autostart + OTLP flush all working, no failed-upload dir. 95 pass, 1 error.

**The one blocking error — `time.non_zero_duration` (an LLM span with `duration=0`, start==end==`1784280998890000000`) — I think is actually reproducible/fixable on this side, not purely upstream in pi.** These spans are built by pilot's OTLP flusher from the extension's JSONL (via `@loongsuite/otel-util-genai`), and every timestamp the extension emits is millisecond-resolution (`BigInt(ms) * 1_000_000n`) — note the failing value ends in six zeros, i.e. exactly `1784280998890 ms × 1e6`. So when a fast step's `llm.request` (`stepStartedAt`) and `llm.response` (`message.timestamp`) land in the same millisecond (or a step emits only the response), start == end → duration 0. Could you confirm this path and stamp the LLM span so `end` is strictly greater than `start` (e.g. clamp duration ≥ 1ns, or use a higher-resolution clock)? That should clear the semantic check without waiting on an upstream pi change.

**Minor (non-blocking):** the extension's `resolveUserId` is effectively overwritten by pilot's daemon-side `configuredUserId`, so it's redundant (harmless).

Overall: strong PR — happy to see it land once the `duration=0`/timestamp path is confirmed. 🎉

**Comment by Goclis** [comment]
@fangxiu-wf Thanks for the live E2E and the precise diagnosis. I confirmed the path and fixed it in `3e8d517`.

The Pi-side root cause had two parts:

- `AssistantMessage.timestamp` is assigned when provider streaming starts, not when `message_end` is delivered. The Extension was incorrectly using it as the LLM response end time.
- Pi timestamps are millisecond-resolution, and a response can also arrive without a preceding `context` event. Either case could produce a response-only or same-millisecond pair, which the downstream converter turns into a zero-duration span.

The fix stays within the Pi Extension:

- use `message_end` receipt time as the LLM completion timestamp;
- emit a minimal fallback `llm.request` if the response has no preceding context event;
- guarantee paired LLM and TOOL events differ by at least 1ms (1ns is insufficient because the converter consumes numeric millisecond timestamps);
- keep `observed_time_unix_nano >= time_unix_nano` after the adjustment.

Regression tests cover same-millisecond LLM pairs, response-only events, actual `message_end` completion timing, and same-millisecond TOOL pairs. They also feed the emitted records through `convertEventLogToReadableSpans` and assert non-zero LLM/TOOL spans.

I also ran a fresh real Pi Coding Agent `0.79.6` E2E using the locally upgraded Pilot build:

- real LLM call plus one real `bash` tool invocation;
- 6 collected event-log records: 2 request, 2 response, 1 tool.call, 1 tool.result;
- 7 exported spans: ENTRY 1, AGENT 1, STEP 2, LLM 2, TOOL 1;
- observed durations: LLM 3243ms / 1647ms, TOOL 8ms;
- all spans had strictly positive duration and valid parent/child timing;
- repository validator: `time.non_zero_duration` PASS, 0 errors (one unrelated non-blocking warning that ENTRY had no input messages).

On the minor `resolveUserId` point: agreed that the daemon's configured user ID is authoritative in the normal Pilot path. I retained the Extension value so the raw JSONL remains self-contained when inspected or consumed independently; no behavioral dependency is placed on it downstream.

Could you please take another look when convenient?

**Comment by fangxiu-wf** [comment]
Hi @Goclis — thank you so much for this contribution! 🙌 First-class Pi Coding Agent instrumentation is exactly the kind of agent coverage we've been hoping for, and honestly this is a really lovely implementation: dependency-free, fail-open, and thoughtfully privacy-aware with the content gating. A pleasure to review.

I re-ran a 1h rolling E2E on `3e8d517` and the data collection quality is excellent 🎉 — 12/12 rounds, 0 errors, the full `ENTRY → AGENT → STEP → LLM → TOOL` tree complete and correct, token summation and tool↔LLM matching all passing. And your timestamp fix fully resolved the earlier `time.non_zero_duration` issue — every LLM/TOOL span now has strictly positive duration. Really nice work.

There are just two small things left before we can merge, and neither is a big lift:

**1. ENTRY / AGENT spans don't yet carry `gen_ai.input.messages`** (warning: `semantic.entry_input_exists → missing input.messages`).

I traced this to root cause: the extension never emits an `event.name: "other"` record. The downstream converter (`@loongsuite/otel-util-genai` `event-log/converter.js`) populates ENTRY/AGENT input *only* from `allUserInputEvents`:

```js
// converter.js — "Extract user-input events that feed ENTRY span's input.messages"
const otherWithMessages = allRecords.filter(r =>
  r["event.name"] === EventName.OTHER &&
  (r["gen_ai.input.messages_delta"] || r["gen_ai.input.messages"]));
const allUserInputEvents = [...otherWithMessages, ...userHooks];
const entryInv = buildEntryInvocation(parentRecords, allUserInputEvents, common);
const agentInv = buildInvokeAgentInvocation(parentRecords, allUserInputEvents, common);
```

Both sources come up empty for Pi:
- **`otherWithMessages`** — the extension only emits `llm.request` / `llm.response` / `tool.call` / `tool.result`, never `event.name: "other"`.
- **`userHooks`** (the deprecated fallback: `llm.request` with *no* step.id, *no* model, *no* paired response) — Pi's `llm.request` always goes through `commonFields`, which sets `gen_ai.step.id` and `gen_ai.request.model`, and it always has a paired `llm.response`, so it's treated as a real LLM request rather than a user-hook.

So `allUserInputEvents` ends up empty and ENTRY/AGENT get no input source. The user prompt *is* captured — but it's attached to the real `llm.request` (`emitLlmRequest`: `record['gen_ai.input.messages'] = messages`), so it only lands on the **LLM** span. That's why the LLM layer looks complete while the two top-level spans are empty.

Suggested fix (matches spec §5 做法 A, same as codex/qoder — e.g. `codex-transcript-builder.ts` emits `event.name: "other"`): emit **one `event.name: "other"` record per turn** carrying `gen_ai.input.messages_delta` = the new user message(s) for that turn. The data is already available in the `context` event's `event.messages` (the trailing `user`/`custom` message). A few notes that might save you some time:
- use the **delta** (only the new user input) rather than the full history, to avoid inflating ENTRY input;
- don't set `gen_ai.step.id` on this record, so it isn't paired as a real LLM step;
- gate it on `captureContent`, consistent with your existing handlers.

**2. CI is currently red.** `build-and-test` fails on Node 20 and 22 (Node 18 passes). The failing test is `tests/unit/utils/pid-utils.test.ts` — not a file this PR touches, and it looks environment-dependent (a `process-scan` liveness check), so it's very likely just flaky rather than a real regression. A re-run may well clear it; if it reproduces, it's worth a quick glance to confirm it's unrelated to your change.

Once those two are sorted we'll be happy to merge this in. Thanks again for the great work — and I'd genuinely love to see you stay involved with more agent integrations down the road! 🚀

**Review by ralf0131 — APPROVED** [review]
### Summary
Re-reviewing commit `3e8d517` ("fix: prevent zero-duration Pi spans") — the contributor added `timestampMillis()` and `timestampStrictlyAfter()` helpers to guarantee end timestamps are strictly after start timestamps, preventing zero-duration spans in the downstream OTel converter. Also adds `requestStartedAt` tracking and a fallback LLM-request emission on `message_end` when no context event preceded it. The new edge-case tests (same-millisecond timing, fallback request, tool-call ordering) all pass.

The code is correct, well-structured, and well-tested. **Approved.**

### CI Status (unrelated failure)
CI fails on Node 20 & 22 matrix jobs, but the failure is in **`tests/unit/utils/pid-utils.test.ts`** (process-liveness detection: `expected { running: false, pid: 123 } to match { running: true, pid: 456 }`) — a file **not touched by this PR**. The pi-coding-agent extension tests all pass (163/164 test files green). This appears to be a pre-existing flaky test on Node 20/22 (passes on Node 18). Recommend re-running CI or investigating `pid-utils.test.ts` separately.

---
*Automated review by github-manager-bot*

File: src/inputs/pi-coding-agent-log/pi-coding-agent-log-input.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  readonly agentType = ClientType.PiCodingAgent;
+
+  constructor(opts?: Partial<HookInputOptions> & { stateStore: HookInputOptions['stateStore'] }) {
+    super({
+      stateStore: opts!.stateStore,
+      logDir: opts?.logDir ?? resolveHome('~/.loongsuite-pilot/logs/pi-coding-agent'),
+      logPrefix: opts?.logPrefix ?? 'pi-coding-agent',
+      pollIntervalMs: opts?.pollIntervalMs ?? 30_000,
+    });
+  }
```
[the remark this answers is not in the document]
Comment (Goclis) [reply src/inputs/pi-coding-agent-log/pi-coding-agent-log-input.ts]: Fixed in `b473ff9`. `PiCodingAgentLogInput` now requires an options object with a mandatory `stateStore`; JavaScript callers that bypass the TypeScript contract receive an explicit `PiCodingAgentLogInput requires a stateStore` error. A regression test covers the missing-dependency path.

File: src/inputs/pi-coding-agent-log/pi-coding-agent-log-input.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+export class PiCodingAgentLogInput extends BaseHookInput {
+  readonly id = 'pi-coding-agent-log';
+  readonly agentType = ClientType.PiCodingAgent;
+
+  constructor(opts?: Partial<HookInputOptions> & { stateStore: HookInputOptions['stateStore'] }) {
+    super({
+      stateStore: opts!.stateStore,
+      logDir: opts?.logDir ?? resolveHome('~/.loongsuite-pilot/logs/pi-coding-agent'),
```
[the remark this answers is not in the document]
Comment (Goclis) [reply src/inputs/pi-coding-agent-log/pi-coding-agent-log-input.ts]: Fixed in `b473ff9`. Added a shared Pi log-directory resolver with the precedence `explicit dataDir` → `LOONGSUITE_PILOT_DATA_DIR` → `~/.loongsuite-pilot`. The constructor, `checkAvailability()`, and `getWatchPaths()` now all use the same resolver. Tests cover both the environment-backed default and an explicitly supplied data directory.

File: src/inputs/pi-coding-agent-log/pi-coding-agent-log-input.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    });
+  }
+
+  static async checkAvailability(): Promise<boolean> {
+    return directoryExists(resolveHome('~/.loongsuite-pilot/logs/pi-coding-agent'));
+  }
+
+  static getWatchPaths(): string[] {
+    return [resolveHome('~/.loongsuite-pilot/logs/pi-coding-agent')];
+  }
```
[the remark this answers is not in the document]
Comment (Goclis) [reply src/inputs/pi-coding-agent-log/pi-coding-agent-log-input.ts]: Fixed in `b473ff9` together with the related constructor/default-path thread. `getWatchPaths(dataDir)` and `checkAvailability(dataDir)` now resolve the same `$PILOT_DATA/logs/pi-coding-agent` location as the Extension and Orchestrator, with regression coverage for explicit and environment-provided data directories.

File: assets/plugins/pi-coding-agent/index.mjs [hunk 9]
```diff
@@ -0,0 +1,510 @@
...
+function appendLogFile(fileName, content) {
+  for (let attempt = 0; attempt < 2; attempt++) {
+    try {
+      ensureLogDir();
+      fs.appendFileSync(path.join(resolveLogDir(), fileName), content, {
+        encoding: 'utf8',
+        mode: 0o600,
+      });
+      return;
+    } catch (error) {
```
[the remark this answers is not in the document]
Comment (Goclis) [reply assets/plugins/pi-coding-agent/index.mjs:131]: Fixed in `b473ff9`. Log appends now use `openSync(..., a, 0o600)`, then `fchmodSync(fd, 0o600)` before writing on non-Windows, with the descriptor closed in `finally`. This covers both JSONL and error logs and also re-tightens the actual inode after runtime log rotation/replacement. Regression tests cover pre-existing permissive files and same-path file replacement.

**Comment by Goclis** [comment]
@fangxiu-wf Thanks again for the detailed E2E diagnosis. The ENTRY/AGENT input gap is fixed in `b473ff9`.

The Pi Extension now emits `event.name: "other"` records carrying only the new trailing user/custom `gen_ai.input.messages_delta` for each Pi turn that has new user input. These records intentionally have no `gen_ai.step.id` or request model, are gated by `captureMessageContent`, and fall back to the `before_agent_start` prompt when an assistant response arrives without a preceding `context` event.

The regression test feeds the emitted records through `convertEventLogToReadableSpans` and verifies that both ENTRY and AGENT receive the expected input messages without duplicating repeated context callbacks. The same commit also addresses the remaining inline review findings around the required `stateStore` contract, standalone `$PILOT_DATA` path resolution, and restrictive permissions for existing or rotated JSONL/error files.

Validation:
- full suite: 153 test files passed, 1707 tests passed, 2 skipped;
- Node 20 Pi-focused tests: 27/27 passed;
- typecheck and build passed.

The previous `pid-utils` CI failure remains unrelated to this PR; this push should trigger a fresh workflow run.

**Comment by Goclis** [comment]
Follow-up: the new CI run for `b473ff9` was created, but GitHub marked it `action_required` because this first-time-contributor workflow needs maintainer approval: https://github.com/alibaba/loongsuite-pilot/actions/runs/29972636556

@fangxiu-wf could you please approve the workflow run and re-review the new commit when convenient?

**Review by ralf0131 — APPROVED** [review]
﻿### Summary
Continues the excellent Pi Coding Agent observability implementation — adds comprehensive plugin, input handler, installer, and test coverage. All previous [Info] suggestions have been addressed across review iterations.

### Overall Assessment
Well-structured, production-ready implementation with proper fail-open design, privacy controls, and comprehensive test coverage.

### Positive Notes
- Clean fail-open design ensures telemetry errors never interrupt agent operation
- Proper serialization with depth/string limits prevents memory issues
- Privacy-conscious content capture toggle (`shouldCaptureContent`)
- Zero-duration span prevention (`timestampStrictlyAfter`) is correctly handled
- 787-line extension test + 150-line input test + 120-line deployment test provides excellent coverage
- Canonical GenAI message format conversion correctly handles text/thinking/toolCall parts
- Proper provider normalization (anthropic/openai/gemini/bedrock/azure/qwen/deepseek)

---
*Automated review by github-manager-bot*

**Comment by fangxiu-wf** [comment]
Hi @Goclis — I re-ran the E2E validation on the latest commit and confirmed that the missing `gen_ai.input.messages` on the ENTRY and AGENT spans has been fully resolved. The complete `ENTRY → AGENT → STEP → LLM → TOOL` pipeline now behaves as expected, and the previously identified issues have all been addressed.

At this point, the implementation is comprehensive, robust, and meets our merge criteria. I’m going to merge the PR now.

Thank you again for the thoughtful work and for responding so carefully to each round of review feedback. We’d be very happy to see you continue contributing new integrations and improvements to LoongSuite Pilot! 🚀
